hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/S3EventsHoodieIncrSource.java

```diff
@@ -172,37 +179,46 @@ public Pair<Option<Dataset<Row>>, String> fetchNextBatch(Option<String> lastCkpt
     String s3FS = props.getString(Config.S3_FS_PREFIX, "s3").toLowerCase();
     String s3Prefix = s3FS + "://";
 
-    // Extract distinct file keys from s3 meta hoodie table
-    final List<Row> cloudMetaDf = source
+    // Create S3 paths
+    final boolean checkExists = props.getBoolean(Config.ENABLE_EXISTS_CHECK, Config.DEFAULT_ENABLE_EXISTS_CHECK);
+    SerializableConfiguration serializableConfiguration = new SerializableConfiguration(sparkContext.hadoopConfiguration());
+    List<String> cloudFiles = source
         .filter(filter)
         .select("s3.bucket.name", "s3.object.key")
         .distinct()
-        .collectAsList();
-    // Create S3 paths
-    final boolean checkExists = props.getBoolean(Config.ENABLE_EXISTS_CHECK, Config.DEFAULT_ENABLE_EXISTS_CHECK);
-    List<String> cloudFiles = new ArrayList<>();
-    for (Row row : cloudMetaDf) {
-      // construct file path, row index 0 refers to bucket and 1 refers to key
-      String bucket = row.getString(0);
-      String filePath = s3Prefix + bucket + "/" + row.getString(1);
-      if (checkExists) {
-        FileSystem fs = FSUtils.getFs(s3Prefix + bucket, sparkSession.sparkContext().hadoopConfiguration());
-        try {
-          if (fs.exists(new Path(filePath))) {
-            cloudFiles.add(filePath);
-          }
-        } catch (IOException e) {
-          LOG.error(String.format("Error while checking path exists for %s ", filePath), e);
-        }
-      } else {
-        cloudFiles.add(filePath);
-      }
-    }
+        .mapPartitions((MapPartitionsFunction<Row, String>)  fileListIterator -> {
+          List<String> cloudFilesPerPartition = new ArrayList<>();
+          final Configuration configuration = serializableConfiguration.newCopy();
+          fileListIterator.forEachRemaining(row -> {
+            String bucket = row.getString(0);
+            String filePath = s3Prefix + bucket + "/" + row.getString(1);
+            String decodeUrl = null;
+            try {
+              decodeUrl = URLDecoder.decode(filePath, StandardCharsets.UTF_8.name());
+              if (checkExists) {
+                FileSystem fs = FSUtils.getFs(s3Prefix + bucket, configuration);
+                if (fs.exists(new Path(decodeUrl))) {
+                  cloudFilesPerPartition.add(decodeUrl);
+                }
+              } else {
+                cloudFilesPerPartition.add(decodeUrl);
+              }
+            } catch (IOException e) {
+              LOG.error(String.format("Error while checking path exists for %s ", decodeUrl), e);
+            } catch (Throwable e) {
+              LOG.warn("Failed to add cloud file ", e);
```

Comment:
didn't realize this before...in the original logic, any exception other than IOException will fail the fetch right? here it'll silence it.

Reply:
@xushiyan changed to throw the exception again.